kernel/src/expressions/mod.rs

```diff
@@ -244,6 +245,82 @@ impl OpaqueExpression {
     }
 }
 
+/// An optional field name. When using a [`Transform`] to insert new fields, this is used to specify
+/// which field, if any, the new fields should be inserted after.
+///
+/// NOTE: This unusual type is necessary because `Option` does not impl `Borrow` and a hash map with
+/// `Option<String>` keys can only be probed with owned strings. In contrast, `Cow` _does_ impl
+/// `Borrow` so we can probe the map with borrowed keys, e.g. `Some(Cow::Borrowed("key"))`.
+pub type FieldNameOpt = Option<Cow<'static, str>>;
+
+/// A transformation that efficiently represents sparse modifications to struct schemas.
+///
+/// `Transform` achieves `O(changes)` space complexity instead of `O(schema_width)` by only
+/// specifying those fields that actually change (inserted, replaced, or deleted). Any input field
+/// not specifically mentioned by the transform is passed through, unmodified and with the same
+/// relative field ordering. This is particularly useful for wide schemas where only a few columns
+/// need to be modified and/or dropped, or where a small number of columns need to be injected.
+#[derive(Debug, Clone, PartialEq, Default)]
+pub struct Transform {
+    /// The path to the nested input struct this transform operates on (if any). If no path is
+    /// given, the transform operates directly on top-level columns.
+    pub input_path: Option<ColumnName>,
+    /// A set of fields names to be replaced, along with their replacement expression (if any). A
+    /// replaced field with no replacement expression will be dropped from the output.
+    pub field_replacements: HashMap<String, Option<ExpressionRef>>,
+    /// A set of new fields to inject, organized by the name of field they should follow (if any).
+    pub field_insertions: HashMap<FieldNameOpt, Vec<ExpressionRef>>,
+}
+
+impl Transform {
+    /// Create a new top-level identity transform. Field transforms and pathing can then be added as
+    /// desired, using the various `with_xxx` helper methods.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Specifies the path to input data this transform operates on. The path must reference a
+    /// struct whose fields will be transformed.
+    pub fn with_input_path<A>(mut self, path: impl IntoIterator<Item = A>) -> Self
+    where
+        ColumnName: FromIterator<A>,
+    {
+        self.input_path = Some(ColumnName::new(path));
+        self
+    }
+
+    /// Specifies a field to drop.
+    pub fn with_dropped_field(mut self, name: impl Into<String>) -> Self {
```

Comment:
I don't see that we use/test `with_dropped_field` nor `with_replaced_field` - do we want to have these or just use underlying hash maps? I suppose they are simple enough but if we decide to keep we may just want to have quick test?